WalletWasabi/Services/LegalChecker.cs

```diff
@@ -0,0 +1,134 @@
+using Nito.AsyncEx;
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Legal;
+using WalletWasabi.Logging;
+using WalletWasabi.Models;
+
+namespace WalletWasabi.Services
+{
+	public class LegalChecker : IDisposable
+	{
+		public const string LegalFolderName = "Legal";
+		public const string ProvisionalLegalFolderName = "Provisional";
+
+		private bool _disposedValue;
+
+		public LegalChecker(string dataDir)
+		{
+			LegalFolder = Path.Combine(dataDir, LegalFolderName);
+			ProvisionalLegalFolder = Path.Combine(LegalFolder, ProvisionalLegalFolderName);
+		}
+
+		public event EventHandler<LegalDocuments>? AgreedChanged;
+
+		public event EventHandler<LegalDocuments>? ProvisionalChanged;
+
+		/// <remarks>Lock object to guard <see cref="CurrentLegalDocument"/> and <see cref="ProvisionalLegalDocument"/> property.</remarks>
+		private AsyncLock LegalDocumentLock { get; } = new();
+
+		private UpdateChecker? UpdateChecker { get; set; }
+		public string LegalFolder { get; }
+		public string ProvisionalLegalFolder { get; }
+		public LegalDocuments? CurrentLegalDocument { get; private set; }
+		private LegalDocuments? ProvisionalLegalDocument { get; set; }
+
+		public async Task InitializeAsync(UpdateChecker updateChecker)
```

Comment:
I would consider not making `LegalChecker` being so tightly dependent on `UpdateChecker`. The callback can be outside of `LegalChecker` and then `LegalChecker` would not need to be disposable.

Reply:
Makes sense in the manner of simplicity. It would break the encapsulation of UpdateChecker.